lang/common/proto/bridge/DriverServiceProtocol.proto

```diff
@@ -0,0 +1,149 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+syntax = "proto3";
+
+option java_generic_services = true;
+option java_multiple_files = true;
+option java_package = "org.apache.reef.bridge.proto";
+option java_outer_classname = "DriverBridgeProtocol";
+option csharp_namespace = "Org.Apache.REEF.Bridge.Proto";
+
+package driverbridge;
+
+import "DriverCommonProtocol.proto";
+
+// The java driver service definition.
+service DriverService {
+  // Driver client registration
+  rpc RegisterDriverClient (DriverClientRegistration) returns (Void) {}
+
+  // Request for resources
+  rpc RequestResources (ResourceRequest) returns (Void) {}
+
+  // Request system shutdown
+  rpc Shutdown (ShutdownRequest) returns (Void) {}
+
+  // Request for setting an alarm
+  rpc SetAlarm (AlarmRequest) returns (Void) {}
+
+  // Request operation on an allocated evaluator
+  rpc AllocatedEvaluatorOp (AllocatedEvaluatorRequest) returns (Void) {}
+
+  // Request operation on an active context
+  rpc ActiveContextOp (ActiveContextRequest) returns (Void) {}
+
+  // Request operation on a running task
+  rpc RunningTaskOp (RunningTaskRequest) returns (Void) {}
+}
+
+message DriverClientRegistration {
+  // The client's host
+  string host = 1;
+
+  // The client's server port
+  int32 port = 2;
+}
+
+// The request message containing resource request.
+message ResourceRequest {
+  repeated string node_name_list = 1;
+
+  repeated string rack_name_list = 2;
+
+  int32 resource_count = 3;
+
+  int32 memory_size = 4;
+
+  int32 priority = 5;
+
+  int32 cores = 6;
+
+  bool relax_locality = 7;
+
+  string runtime_name = 8;
+}
+
+// Request for an alarm to be set
+message AlarmRequest {
```

Comment:
I was concerned about an "idle driver" (on the java side) if all the client does (at a moment) is schedule an alarm. Thoughts?

Reply:
We have `org.apache.reef.runtime.common.driver.idle.DriverIdleManager` to mediate the problem of premature Driver shutdown.  It is invoked by the `Clock` whenever it looks as if the Driver could be idle. The `DriverIdleManager` then consoles a bunch of `org.apache.reef.runtime.common.driver.idle.DriverIdlenessSource` instances bound to `org.apache.reef.driver.parameters.DriverIdleSources` to determine whether the Driver is, in fact, idle.

One could use this mechanism to avoid the premature shutdown here as well: Implement a `DriverIdlenessSource` that bases its idleness assessment on the state of the side process. Would that alleviate the need for cross-process alarms?

Reply:
Let me make sure I understand your suggestion:
1. on the java side, monitor the exit value of the client process.
2. if exited, then the driver is idle
3. otherwise the driver is not idle
When you say "state of the side process", I'm assuming the exit value since that's the only state we know about.

If that is what you mean, then that would require an idle check on the client process that exits when idle. That works, but it places a burden on the client driver harness author (not the application) to develop such an idle check that matches the semantics of the java side i.e., no evalautors or alarms, currently. As I am currently implementing such a harness in java, that would require me to maintain state, which I currently do not forsee needing i.e., client process harness is simply a protocol buffer from/to object translator and dispatcher. 

Given that applications tend not to set many alarms (IMO), I would lean towards the cross-process alarm approach.

Reply:
> Given that applications tend not to set many alarms (IMO), I would lean towards the cross-process alarm approach.

Is this only for application alarms or also for the functioning of the Driver-Driver system? I'd imagine we could run into an idle situation after firing `DriverStart`, if the client Driver doesn't respond fast enough. How do we guard against that? Via Alarms? If so, how do we choose the timeout? What other situations do we need to guard with Alarms?

Reply:
Only for application alarms. 

I’m thinking that driver start (and only it) will be synchronous, such that on return, evaluators have been requested and/or alarms have been set.  

I also think we need to respect the order of actions taken by the application eg if the application requests evaluation before closing the last active evaluators then the real (java) driver should see that same order.

Reply:
> Only for application alarms.
> I’m thinking that driver start (and only it) will be synchronous, 

What if there is a `CompletedEvaluator` event for the last Evaluator? Don't we run into the same potential time-out? Would we have to make that synchronous as well?

Reply:
Possibly, but then I wonder about how this happens in the base java driver since such events are dispatched asynchronously on a thread pool. I’ll have to look into replicating such logic or implementing an idleness checker that waits for a period of time

Reply:
> I suggest keeping it simple by setting an idle checker that waits a minute (literally) before declaring things are truly idle. 

Hmm, that would mean that a failing Driver needs at least 1 minute till it properly fails, right? I'm ok with that as a first cut, but we should file a JIRA to remind us that we skipped the more difficult route here.

> Then we implement the entire bridge asynchronously and forget about it; just like the current dispatcher java driver as far as I can tell.

The current Java code doesn't do it like that. It inquires with the thread pools when idleness is fired to confirm that no user code is running on any of those threads.

Reply:
> The current Java code doesn't do it like that. It inquires with the thread pools when idleness is fired to confirm that no user code is running on any of those threads.

I was not aware and it makes sense. How about we do something similar with the Java bridge by having a single synchronous call for idleness. On the driver application side, we can check the thread pool for activity. I can also imagine something similar in C# that checks for outstanding Task activity.

Reply:
Actually, I don’t think such a check needs to be synchronized. How about we fire an async event to the application driver saying: hey we think you’re idle; we (the java driver) are not going to act on it (ie shitdown) unless you (asynchronously) tell us to do so.

Reply:
That sounds like a plan. We still have to guard against the client Driver becoming unresponsive due to bugs and such, though.

Reply:
Sounds good.